### ISSUE TITLE
Improve add widget UI on Flows 

### ISSUE THREAD

**Issue by peterkeating** [OrchardCMS/OrchardCore#3558]
The current implementation of adding widgets within a flow doesn't extend well when there are a large amount of widgets, as shown below.

![flows-existing-add-widget](https://user-images.githubusercontent.com/651190/57046417-4c0f9900-6c68-11e9-85c9-1bf784517c44.png)

An alternative is to replace the dropdown button with a modal popup that'll open the door to other possibilities like filtering, ordering, grouping (grouping may need some more thought behind it), etc... 

I've had a quick play with the current implementation and threw together a short screen recording to give a better idea of the direction I was thinking.

![flows-demo-add-widget](https://user-images.githubusercontent.com/651190/57046483-9729ac00-6c68-11e9-9008-ca79d6a0706b.gif)

Be great to get any thoughts/further suggestions and confirmation to whether it's worth implementing something more concrete and submit a PR.

**Comment by sebastienros** [comment]
Next, grouping and custom shapes which would allow us to use custom rendering (icons or images for instance). Search box also maybe?

**Comment by sebastienros** [comment]
The group could be a type setting, which could later be integrated in the content type editors if we want to change it.

**Comment by peterkeating** [comment]
@sebastienros Awesome, I'll crack on then! 

Could you explain how you see custom shapes working? Do you mean a shape for customising the display of card that represents the widget in the modal?

**Comment by sebastienros** [comment]
Yes for the shapes. The idea is that you render a widget with a specific display type, like Thumbnail (look at how it's done for Workflows, or Recipe Steps, ...) and then there is a default template named Widget.Thumbnail.cshtml, but because you add a custom alternate like `Widget_Thumnbail-[WithType]` any module could then provide a custom template for each widget. But at the same time it can be done later, AND it might not be that useful as most widgets will be dynamic.

**Comment by peterkeating** [comment]
More often than not we’re creating widgets via recipes, so perhaps the thumbnail could be specified in the content type settings similar to the group?

Still think having the shape would be useful too for consistency and to give the option to override it with an alternate like you’ve mentioned. This would also be useful if you want to override the shape in a custom admin theme too.
